### PR TITLE
Add conv2d transpose NNAPI converter

### PR DESCRIPTION
**Pull request by axitkhurana**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61025
* #61024
* #61023
* #61022
* #61021
* **#59529**
* #59434

Summary:

* Conv2d transpose support
* Quantize WIP

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28926335](https://our.internmc.facebook.com/intern/diff/D28926335)

### CODE REVIEW

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

File: torch/backends/_nnapi/serializer.py [hunk 1]
```diff
@@ -1658,10 +1658,13 @@ def add_qlinear(self, node):
...
 
-    def get_optional_bias(self, jit_bias, weight_tensor):
+    def get_optional_bias(self, jit_bias, weight_tensor, transpose=False):
         ctype, value = self.get_constant_value(jit_bias)
         if ctype.kind() == "NoneType":
-            nnapi_bias_tensor = torch.zeros(weight_tensor.size()[0], dtype=weight_tensor.dtype)
+            if transpose:
+                nnapi_bias_tensor = torch.zeros(weight_tensor.size()[1], dtype=weight_tensor.dtype)
+            else:
+                nnapi_bias_tensor = torch.zeros(weight_tensor.size()[0], dtype=weight_tensor.dtype)
```
Comment (dreiss) [inline torch/backends/_nnapi/serializer.py:1667]:
```
bias_idx = 1 if transpose else 0
nnapi_bias_tensor = torch.zeros(weight_tensor.size()[bias_idx], dtype=weight_tensor.dtype)
```

File: test/test_nnapi.py [hunk 2]
```diff
@@ -503,6 +503,62 @@ def test_conv2d(self):
...
+    def test_conv2d_transpose(self):
+        weight = torch.randn((128, 256, 2, 2))
+        input_dim = (30, 128, 7, 7)
+        inp = torch.randn(input_dim)
+
+        convert_dims = input_dim[:2] + (0, 0)
+        convert_arg = torch.zeros(*convert_dims)
+        convert_args = dict(self.float_and_quant_and_nhwc(convert_arg, 0.03, 128))
+
+        class ConvTransposeMod(torch.nn.Module):
```
Comment (dreiss) [inline test/test_nnapi.py:515]: Can you use `torch.nn.ConvTranspose2d`?

File: test/test_nnapi.py [hunk 3]
```diff
@@ -503,6 +503,62 @@ def test_conv2d(self):
                         limit=limit
                     )
 
+    def test_conv2d_transpose(self):
+        weight = torch.randn((128, 256, 2, 2))
+        input_dim = (30, 128, 7, 7)
```
Comment (dreiss) [inline test/test_nnapi.py:508]: Do we need the data to be this large?

**Review by dreiss — APPROVED** [review]

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

File: test/test_nnapi.py (hunk 3 above)
Comment (axitkhurana) [reply test/test_nnapi.py:508]: Definitely not, was dealing with that size during conversion ended up using it in the test too

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Comment by axitkhurana** [comment]
@axitkhurana has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28926335).

**Pull request closed without merging**